libs/common/src/platform/services/cryptography/bulk-encrypt.service.implementation.ts

```
@@ -0,0 +1,159 @@
+import { firstValueFrom, fromEvent, filter, map, takeUntil, defaultIfEmpty, Subject } from "rxjs";
+import { Jsonify } from "type-fest";
+
+import { BulkEncryptService } from "../../abstractions/bulk-encrypt.service";
+import { CryptoFunctionService } from "../../abstractions/crypto-function.service";
+import { LogService } from "../../abstractions/log.service";
+import { Decryptable } from "../../interfaces/decryptable.interface";
+import { InitializerMetadata } from "../../interfaces/initializer-metadata.interface";
+import { Utils } from "../../misc/utils";
+import { SymmetricCryptoKey } from "../../models/domain/symmetric-crypto-key";
+
+import { getClassInitializer } from "./get-class-initializer";
+
+// TTL (time to live) is not strictly required but avoids tying up memory resources if inactive
+const workerTTL = 60000; // 1 minute
+const maxWorkers = 8;
+const minNumberOfItemsForMultithreading = 400;
+
+export class BulkEncryptServiceImplementation implements BulkEncryptService {
+  private workers: Worker[] = [];
+  private timeout: any;
+
+  private clear$ = new Subject<void>();
+
+  constructor(
+    protected cryptoFunctionService: CryptoFunctionService,
+    protected logService: LogService,
+  ) {}
+
+  /**
+   * Decrypts items using a web worker if the environment supports it.
+   * Will fall back to the main thread if the window object is not available.
+   */
+  async decryptItems<T extends InitializerMetadata>(
+    items: Decryptable<T>[],
+    key: SymmetricCryptoKey,
```

non-blocking: This is technically Platforms code, so really up to @justindbaur, but I'd love to see us check the `key` here as well. To be fair this may not be realistic considering where this is used, so feel free to ignore if so.

---

I would like that if we can make it throw when given `null`. I think I also want it to throw if `items` was null instead of returning `[]` but I see that is how it already works so it's fine to leave for now.

---

I had a look at the corresponding code for the single threaded encrypt service implementation and it only checks for whether the key is null, so I added the same check to the bulk encrypt service. Anything more than that (comparing to private key) should be done in higher layers (when setting the key in state f.e) in my opinion.